### ISSUE TITLE
"irrefutable_let_patterns" shouldn't fire for leading if-let guards

### ISSUE THREAD

**Issue by goffrie** [rust-lang/rust#98361]
[sample code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d80dbfccf6b1054d2930900ac1615d20):

```rust
#![feature(if_let_guard, let_chains)]

pub fn example(x: Option<Option<&str>>) {
    match x {
        Some(y) if let z = y.expect("can't be Some(None)") && z != "foo" => {
            dbg!(z);
        }
        _ => (),
    }
}
```

yields this warning:
```
warning: leading irrefutable pattern in let chain
 --> src/lib.rs:5:20
  |
5 |         Some(y) if let z = y.expect("can't be Some(None)") && z != "foo" => {
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(irrefutable_let_patterns)]` on by default
  = note: this pattern will always match
  = help: consider moving it outside of the construct
```

However, the suggestion to move it out of the construct doesn't make sense, as `y` isn't bound until we've entered the match arm in the first place.

**Comment by est31** [comment]
I've made a PR to turn off the lint in this instance: #103031

The ideal fix would be to check for usages in the prefix of the bindings created by the match, but this is better than the status quo as for lints, a true negative is better than a false positive.
